service/internal/asg/asg.go

```diff
@@ -0,0 +1,297 @@
+package asg
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/to"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/aws-operator/pkg/awstags"
+	"github.com/giantswarm/aws-operator/service/controller/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/key"
+	"github.com/giantswarm/aws-operator/service/internal/asg/internal/cache"
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+
+	Stack        string
+	TagKey       string
+	TagValueFunc func(cr key.LabelsGetter) string
+}
+
+type ASG struct {
+	k8sClient k8sclient.Interface
+
+	asgsCache      *cache.ASGs
+	instancesCache *cache.Instances
+
+	stack        string
+	tagKey       string
+	tagValueFunc func(cr key.LabelsGetter) string
+}
+
+func New(config Config) (*ASG, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+
+	if config.Stack == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Stack must not be empty", config)
+	}
+	if config.TagKey == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.TagKey must not be empty", config)
+	}
+	if config.TagValueFunc == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.TagValueFunc must not be empty", config)
+	}
+
+	a := &ASG{
+		k8sClient: config.K8sClient,
+
+		asgsCache:      cache.NewASGs(),
+		instancesCache: cache.NewInstances(),
+
+		stack:        config.Stack,
+		tagKey:       config.TagKey,
+		tagValueFunc: config.TagValueFunc,
+	}
+
+	return a, nil
+}
+
+func (a *ASG) Drainable(ctx context.Context, obj interface{}) (string, error) {
```

Comment:
@calvix can you check if this implementation makes sense to you? I would use the ASG service in the drainer handlers and configure them according to their controllers.

Reply:
@giantswarm/team-firecracker-pe somebody?

Reply:
Make sense to me 🙂, if I got the logic right we simply try to get from all asgs from ec2 instances which are relevant and then we look at the asgs which are safe to drain by looking at the lifecycle state. 
Still thinking if we may need to wait a bit longer TerminatingProceed or even Terminated state 🤔

Reply:
No sorry because of the lifecycle hook from the asg

Reply:
I was also thinking about some timing based safety net but then I figured we should stick to an event based approach, which is almost always better. The safety net is implemented so that it only picks the next ASG in case the other ASGs are fine and have at least one healthy instance running. So that means 3 out of 3 ASGs have instances in service and then picking one to drain is safe.

Reply:
LGTM